Implement CSS text-transform with multiple values

mdubet opened the pull request:
#### 8c7c9a791532f591a79b3caba4e2231f24873a7d
<pre>
Implement CSS text-transform with multiple values
<a href="https://bugs.webkit.org/show_bug.cgi?id=252157">https://bugs.webkit.org/show_bug.cgi?id=252157</a>
rdar://105381249

Reviewed by Tim Nguyen.

The text-transform CSS property supports having multiples
values with a specific order of operation.

<a href="https://w3c.github.io/csswg-drafts/css-text/#text-transform-property">https://w3c.github.io/csswg-drafts/css-text/#text-transform-property</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-transform-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-transform-valid-expected.txt:
* LayoutTests/inspector/css/getSupportedCSSProperties-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::renderTextTransformFlagsToCSSValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeTextTransform):
(WebCore::CSSPropertyParserHelpers::consumeTextIndent):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::styleDidChange):
(WebCore::applyTextTransform):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::textTransform const):
(WebCore::RenderStyle::setTextTransform):
(WebCore::RenderStyle::initialTextTransform):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextTransform):

Canonical link: <a href="https://commits.webkit.org/261419@main">https://commits.webkit.org/261419@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cc3bb98532a0f843761f77d1ef1ee0d0e4a07fbc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111671 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20807 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/283 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3426 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/120407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115732 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22162 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11887 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117437 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99605 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104571 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/184 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/45359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13286 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/181 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/87793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13776 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19221 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52179 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7949 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15766 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -5556,6 +5556,50 @@ RefPtr<CSSValue> consumeSize(CSSParserTokenRange& range, CSSParserMode mode)
+        if (ident->valueID() == CSSValueLowercase && !uppercase && !capitalize && !lowercase) {
+            lowercase = true;
+            continue;
+        }
```
Can we deduplicate these checks:
```
if (!uppercase && !capitalize && !lowercase) {
   ...
}
```

mdubet replied:
Changed

nt1m reviewed (approved):
r=me with tests rebaselined

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -2772,7 +2772,6 @@ imported/w3c/web-platform-tests/css/css-text/text-transform/text-transform-fullw
 imported/w3c/web-platform-tests/css/css-text/text-transform/text-transform-fullwidth-008.html [ ImageOnlyFailure ]
 imported/w3c/web-platform-tests/css/css-text/text-transform/text-transform-fullwidth-009.html [ ImageOnlyFailure ]
 
-webkit.org/b/143153 imported/w3c/web-platform-tests/css/css-text/text-transform/text-transform-multiple-001.html [ ImageOnlyFailure ]
```
This test needs to be skipped on glib (gtk + wpe) and wincairo where some of this isn't implemented.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -5556,6 +5556,64 @@ RefPtr<CSSValue> consumeSize(CSSParserTokenRange& range, CSSParserMode mode)
+    bool fullWidth = { false };
+    bool uppercase = { false };
+    bool capitalize = { false };
+    bool lowercase  = { false };
```
Can we remove the brackets? `bool fullSizeKana = false;`

nt1m replied:
Also `lowercase` has an extra space before the `=`

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -5556,6 +5556,64 @@ RefPtr<CSSValue> consumeSize(CSSParserTokenRange& range, CSSParserMode mode)
+            fullWidth = true;
+            continue;
+        }
+        bool alreadySet = uppercase || capitalize || lowercase;
```
`foundCase` might be a nicer name.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -5556,6 +5556,64 @@ RefPtr<CSSValue> consumeSize(CSSParserTokenRange& range, CSSParserMode mode)
+RefPtr<CSSValue> consumeTextTransform(CSSParserTokenRange& range)
+{
+    if (consumeIdentRaw<CSSValueNone>(range))
+        return CSSValueList::createSpaceSeparated(CSSPrimitiveValue::create(CSSValueNone));
```
You could return `CSSPrimitiveValue::create(CSSValueNone)` directly without putting it in a list.

mdubet committed:
Implement CSS text-transform with multiple values
https://bugs.webkit.org/show_bug.cgi?id=252157
rdar://105381249

Reviewed by Tim Nguyen.

The text-transform CSS property supports having multiples
values with a specific order of operation.

https://w3c.github.io/csswg-drafts/css-text/#text-transform-property

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-transform-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-transform-valid-expected.txt:
* LayoutTests/inspector/css/getSupportedCSSProperties-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::renderTextTransformFlagsToCSSValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeTextTransform):
(WebCore::CSSPropertyParserHelpers::consumeTextIndent):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::styleDidChange):
(WebCore::applyTextTransform):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::textTransform const):
(WebCore::RenderStyle::setTextTransform):
(WebCore::RenderStyle::initialTextTransform):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextTransform):

Canonical link: https://commits.webkit.org/261419@main
